### PR TITLE
[FLINK-22011][table-runtime-blink] Support local global optimization for window aggregation in runtime

### PR DESCRIPTION
**Pull request by wuchong**
## What is the purpose of the change

Support local global optimization for window aggregation in runtime.

## Brief change log

- Introduced local window agg operator which combines records into acc (`LocalSlicingWindowAggOperator` and `LocalAggRecordsCombiner`)
- Introduced global window agg combiner, i.e. `GlobalAggAccCombiner`.

## Verifying this change

We add parameterized one-phase and two-phase for `WindowAggregateITCase`, and enable default two-phase for `WindowRankITCase` and `WindowDistinctAggregateITCase` and `WindowAggregateJsonITCase`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/LocalSlicingWindowAggOperator.java [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+                        localCombiner);
+    }
+
+    @Override
+    public void processElement(StreamRecord<RowData> element) throws Exception {
+        RowData inputRow = element.getValue();
+        RowData key = keySelector.getKey(inputRow);
+        long sliceEnd = sliceAssigner.assignSliceEnd(inputRow, CLOCK_SERVICE);
+        // may flush to output if buffer is full
+        windowBuffer.addElement(key, sliceEnd, inputRow);
```
Comment (beyond1920) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/LocalSlicingWindowAggOperator.java:123]: Currently, we buffer input record directly. We could also buffer acc which may cost less memory. What do you think?

**Review by beyond1920 — COMMENTED** [review]
@wuchong Thanks for great work. The pull request looks good to me, I only left a few comment.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/LocalSlicingWindowAggOperator.java (hunk 1 above)
Comment (wuchong) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/LocalSlicingWindowAggOperator.java:123]: Use memory buffer can help to reduce GC problems and avoid OOM.

Comment (beyond1920) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/LocalSlicingWindowAggOperator.java:123]: yes, I agree with you at this point. However it does not conflict with buffer what. We could also buffer acc, right?

**Comment by YesOrNo828** [comment]
@wuchong  hello jark, I get a expception by below SQL, and find out : the local and global window aggregate operator support insert-only in input right now. Why ? There is any limition in this case?

`SELECT count(0), sum(score),window_start,window_end, window_time FROM TABLE(
    CUMULATE(TABLE UserScores_event, DESCRIPTOR(tt), INTERVAL '10' SECONDS, INTERVAL '60' SECONDS))
 group by window_start,window_end, window_time`

**Pull request closed without merging**